File: src/whylogs/v2/core/statistics/numbertracker.py
"""
TODO:
    * Implement histograms
"""
import logging
import numbers

import datasketches
import pandas as pd

from whylogs.core.statistics.datatypes import FloatTracker, IntTracker, VarianceTracker
from whylogs.core.statistics.thetasketch import ThetaSketch
from whylogs.core.summaryconverters import histogram_from_sketch, quantiles_from_sketch
from whylogs.proto import NumbersMessage, NumberSummary
from whylogs.util import dsketch, stats

# Parameter controlling histogram accuracy.  Larger = more accurate
DEFAULT_HIST_K = 256

logger = logging.getLogger(__name__)


class NumberTracker:
    """
    Class to track statistics for numeric data.

    Parameters
    ----------
    variance
        Tracker to follow the variance
    floats
        Float tracker for tracking all floats
    ints
        Integer tracker

    Attributes
    ----------
    variance
        See above
    floats
        See above
    ints
        See above
    theta_sketch : `whylabs.logs.core.statistics.thetasketch.ThetaSketch`
        Sketch which tracks approximate cardinality
    """

    def __init__(
        self,
        variance: VarianceTracker = None,
        floats: FloatTracker = None,
        ints: IntTracker = None,
        theta_sketch: ThetaSketch = None,
        histogram: datasketches.kll_floats_sketch = None,
    ):
        # Our own trackers
        if variance is None:
            variance = VarianceTracker()
        if floats is None:
            floats = FloatTracker()
        if ints is None:
            ints = IntTracker()
        if theta_sketch is None:
            theta_sketch = ThetaSketch()
        if histogram is None:
            histogram = datasketches.kll_floats_sketch(DEFAULT_HIST_K)
        self.variance = variance
        self.floats = floats
        self.ints = ints
        self.theta_sketch = theta_sketch
        self.histogram = histogram

    @property
    def count(self):
        return self.variance.count

    def track(self, number):
        """
        Add a number to statistics tracking

        Parameters
        ----------
        number : int, float
            A numeric value
        """
        if pd.isnull(number) or (not isinstance(number, numbers.Real)) or isinstance(number, bool):
            # XXX: this type checking may still be fragile in python.
            return
        self.variance.update(number)
        self.theta_sketch.update(number)
        # TODO: histogram update
        # Update floats/ints counting
        f_value = float(number)
        self.histogram.update(f_value)
        if self.floats.count > 0:
            self.floats.update(f_value)
        # Note: this type checking is fragile in python.  May want to include
        # numpy.integer in the type check
        elif isinstance(number, int):
            self.ints.update(number)
        else:
            self.floats.add_integers(self.ints)
            self.ints.set_defaults()
            self.floats.update(f_value)

    def merge(self, other):
        # Make a copy of the histogram
        hist_copy = datasketches.kll_floats_sketch.deserialize(self.histogram.serialize())
        hist_copy.merge(other.histogram)

        theta_sketch = self.theta_sketch.merge(other.theta_sketch)
        return NumberTracker(
            variance=self.variance.merge(other.variance),
            floats=self.floats.merge(other.floats),
            ints=self.ints.merge(other.ints),
            theta_sketch=theta_sketch,
            histogram=hist_copy,
        )

    def to_protobuf(self):
        """
        Return the object serialized as a protobuf message
        """
        opts = dict(
            variance=self.variance.to_protobuf(),
            compact_theta=self.theta_sketch.serialize(),
            histogram=self.histogram.serialize(),
        )
        if self.floats.count > 0:
            opts["doubles"] = self.floats.to_protobuf()
        elif self.ints.count > 0:
            opts["longs"] = self.ints.to_protobuf()
        msg = NumbersMessage(**opts)
        return msg

    @staticmethod
    def from_protobuf(message: NumbersMessage):
        """
        Load from a protobuf message

        Returns
        -------
        number_tracker : NumberTracker
        """
        theta = None
        if message.compact_theta is not None and len(message.compact_theta) > 0:
            theta = ThetaSketch.deserialize(message.compact_theta)
        elif message.theta is not None and len(message.theta) > 0:
            logger.warning("Possible missing data. Non-compact theta sketches are no longer supported")

        opts = dict(
            theta_sketch=theta,
            variance=VarianceTracker.from_protobuf(message.variance),
            histogram=dsketch.deserialize_kll_floats_sketch(message.histogram),
        )
        if message.HasField("doubles"):
            opts["floats"] = FloatTracker.from_protobuf(message.doubles)
        if message.HasField("longs"):
            opts["ints"] = IntTracker.from_protobuf(message.longs)
        return NumberTracker(**opts)

    def to_summary(self):
        """
        Construct a `NumberSummary` message

        Returns
        -------
        summary : NumberSummary
            Summary of the tracker statistics
        """
        if self.variance.count == 0:
            return

        stddev = self.variance.stddev()
        doubles = self.floats.to_protobuf()
        if doubles.count > 0:
            mean = self.floats.mean()
            min = doubles.min
            max = doubles.max
        else:
            mean = self.ints.mean()
            min = float(self.ints.min)
            max = float(self.ints.max)

        unique_count = self.theta_sketch.to_summary()
        histogram = histogram_from_sketch(self.histogram)
        quant = quantiles_from_sketch(self.histogram)
        num_records = self.variance.count
        cardinality = unique_count.estimate
        if doubles.count > 0:
            discrete = False
        else:
            discrete = stats.is_discrete(num_records, cardinality)

        return NumberSummary(
            count=self.variance.count,
            stddev=stddev,
            min=min,
            max=max,
            mean=mean,
            histogram=histogram,
            quantiles=quant,
            unique_count=unique_count,
            is_discrete=discrete,
        )

File: java/core/src/main/java/com/whylogs/core/statistics/NumberTracker.java
package com.whylogs.core.statistics;

import com.google.protobuf.ByteString;
import com.whylogs.core.message.NumbersMessage;
import com.whylogs.core.statistics.datatypes.DoubleTracker;
import com.whylogs.core.statistics.datatypes.LongTracker;
import com.whylogs.core.statistics.datatypes.VarianceTracker;
import com.whylogs.core.utils.sketches.ThetaSketch;
import java.util.Optional;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.val;
import org.apache.datasketches.kll.KllFloatsSketch;
import org.apache.datasketches.memory.Memory;
import org.apache.datasketches.theta.Union;

@Getter
@Builder(setterPrefix = "set")
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class NumberTracker {
  // our own trackers
  VarianceTracker variance;
  DoubleTracker doubles;
  LongTracker longs;

  // sketches
  KllFloatsSketch histogram; // histogram
  Union thetaSketch; // theta sketch for cardinality

  public NumberTracker() {
    this.variance = new VarianceTracker();
    this.doubles = new DoubleTracker();
    this.longs = new LongTracker();

    this.histogram = new KllFloatsSketch(256);
    this.thetaSketch = Union.builder().buildUnion();
  }

  public void track(Number number) {
    double dValue = number.doubleValue();
    variance.update(dValue);
    histogram.update((float) dValue);
    thetaSketch.update(dValue);

    if (doubles.getCount() > 0) {
      doubles.update(dValue);
    } else if (number instanceof Long || number instanceof Integer) {
      longs.update(number.longValue());
    } else {
      doubles.addLongs(longs);
      longs.reset();
      doubles.update(dValue);
    }
  }

  public void add(NumberTracker other) {
    if (other == null) {
      return;
    }

    this.variance.add(other.variance);
    this.doubles.add(other.doubles);
    this.longs.add(other.longs);
    this.histogram.merge(other.histogram);
  }

  @NonNull
  public NumberTracker merge(NumberTracker other) {
    if (other == null) {
      return this;
    }

    val unionHistogram = KllFloatsSketch.heapify(Memory.wrap(this.histogram.toByteArray()));
    unionHistogram.merge(other.histogram);

    val thetaUnion = Union.builder().buildUnion();
    thetaUnion.update(this.thetaSketch.getResult());
    thetaUnion.update(other.thetaSketch.getResult());

    return NumberTracker.builder()
        .setVariance(this.variance.merge(other.variance))
        .setDoubles(this.doubles.merge(other.doubles))
        .setLongs(this.longs.merge(other.longs))
        .setThetaSketch(thetaUnion)
        .setHistogram(unionHistogram)
        .build();
  }

  public NumbersMessage.Builder toProtobuf() {
    val builder =
        NumbersMessage.newBuilder()
            .setVariance(variance.toProtobuf())
            .setHistogram(ByteString.copyFrom(histogram.toByteArray()));

    if (this.doubles.getCount() > 0) {
      builder.setDoubles(this.doubles.toProtobuf());
    } else if (this.longs.getCount() > 0) {
      builder.setLongs(this.longs.toProtobuf());
    }

    builder.setCompactTheta(ThetaSketch.serialize(thetaSketch));

    return builder;
  }

  @NonNull
  public static NumberTracker fromProtobuf(NumbersMessage message) {
    val hMem = Memory.wrap(message.getHistogram().toByteArray());
    val builder =
        NumberTracker.builder()
            .setHistogram(KllFloatsSketch.heapify(hMem))
            .setVariance(VarianceTracker.fromProtobuf(message.getVariance()));

    Optional.ofNullable(message.getDoubles())
        .map(DoubleTracker::fromProtobuf)
        .ifPresent(builder::setDoubles);
    Optional.ofNullable(message.getLongs())
        .map(LongTracker::fromProtobuf)
        .ifPresent(builder::setLongs);

    builder.setThetaSketch(ThetaSketch.deserialize(message.getCompactTheta()));

    return builder.build();
  }
}
